### PR TITLE
Prevent creation of on-chain invoices below the dust limit

### LINKED ISSUE #3071: Prevent the creation of below dust limit on-chain payment method for small amount invoices
**Describe the bug**
Currently a user can create an invoice for a small amount i.e. 0.10 EUR and get back a qr code for an onchain payment that is below bitcoin's dust limit. Thus no customer could actually pay such an invoice from there wallet.

Imho it would be better to skip the creation of btc-onchain payment method in such cases, so that the invoice view moves on the lightning payment option (if it is available) instantly.

**To Reproduce the bug**
Steps to reproduce the reported bug:
1. create invoice with amout 0.10 EUR
2. try to pay it with your wallet (spoiler: it won't let you)

**Expected behavior**

As described above: I wouldn't allow onchain payment method creation in such cases.

**Your BTCPay Environment (please complete the following information):**
- BTCPay Server Version:    v1.3.2.0 (altcoins)
- Deployment Method:        Docker
- Browser:                  Chrome

### PR DESCRIPTION
**Pull request by satwo**
Fixes #3071.

### CODE REVIEW

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs [hunk 1]
```diff
@@ -190,6 +190,18 @@ private string GetPaymentMethodName(BTCPayNetworkBase network)
             }
 
             var reserved = await prepare.ReserveAddress;
+
+            var txOut = network.NBitcoinNetwork.Consensus.ConsensusFactory.CreateTxOut();
+            txOut.ScriptPubKey = reserved.Address.ScriptPubKey;
+            var dust = txOut.GetDustThreshold();
+            var amount = paymentMethod.Calculate().Due.GetValue(network);
```
Comment (NicolasDorier) [inline BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs:197]:
```suggestion
-            var amount = paymentMethod.Calculate().Due.GetValue(network);
+           var amount = paymentMethod.Calculate().Due;
```

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs [hunk 2]
```diff
@@ -190,6 +190,18 @@ private string GetPaymentMethodName(BTCPayNetworkBase network)
...
+            
+            if (amount < dust.ToDecimal(MoneyUnit.BTC))
+            {
+                throw new NotSupportedException("Amount below the dust threshold. For amounts of this size, it is recommended to enable " +
```
[the remark this answers is not in the document]
Comment (satwo) [reply BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs:201]: I was going back and forth between these two :) I agree this is a better choice.

[the remark this answers is not in the document]
Comment (satwo) [reply BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs:201]: I was going back and forth between these two :) I agree this is a better choice.

[the remark this answers is not in the document]
Comment (satwo) [reply BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs]: That's a much clearer error message, too (no need for a stack trace here).

**Comment by satwo** [comment]
Will fix the tests.

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs [hunk 3]
```diff
@@ -196,7 +196,7 @@ private string GetPaymentMethodName(BTCPayNetworkBase network)
             var dust = txOut.GetDustThreshold();
             var amount = paymentMethod.Calculate().Due;
             
-            if (amount < dust)
+            if (amount < dust && paymentMethod.ParentEntity.Type != InvoiceType.TopUp)
```
Comment (satwo) [inline BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs:199]: @NicolasDorier Would it be better to use here
`&& paymentMethod.ParentEntity.Type == InvoiceType.Standard`  ?

Comment (NicolasDorier) [reply BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs:199]: seems ok

**Comment by NicolasDorier** [comment]
Fixed the test, also extended the test to actually test the dust limit.

**Pull request merged**